### DISCUSSION TITLE
High Server load

### DISCUSSION THREAD

**Discussion by Crimson1110** [Security-Onion-Solutions/securityonion#12075] [Unsupported Versions] [answered]
Hello community,

I am a newby for Security Onion. I am testing Security Onion 2.4.30 under single Proxmox host with 2x E5-2430L v3 (20M Cache, 1.80 GHz), 256Gb RAM and Raid5  SSD Crucial MX500  with 5 disks. The Manager Node  & Search node has 12 cores  / 64 Gb Ram.  Both have java heapsize of 16Gb (InfluxDB shows the use percent flows under 12Gb).  Logstash has 4Gb of heap. The sensor node  has 6 cores / 16Gb.  All have 400Gb of hardisk space. 

I added to fleet around 20 linux machines which I use for hosting my personal websites / mail / Proxmox etc.  The linux VMs other than SO are very idle in terms of load and network. InfluxDB shows iodelay as 0.0%. Playbook is on.

I did different combinations of heap size, CPU, total memory etc.  but I am not able to lower the server load increasing to 16-18 within 2 days after I turn on the SO VMs .  The SO VMs seems to utilize only 50-60% CPU on average. 

What am I missing? 

Thank you in advance.

**Comment by dougburks** [comment]
> I am not able to lower the server load increasing to 16-18 within 2 days after I turn on the SO VMs

Which server load are you referring to? The Proxmox host?

**Reply by Crimson1110** [reply]
Yes, sorry if I was not clear, the Proxmox host has a high load with SO. It seems the cause is the Elasticsearch but I don't understand why.  Without SO the load is around 0.2 - 0.5. With SO after it deploys the Elasticsearch container goes up to 18 gradualy in 2-3 days.

**Reply by dougburks** [reply]
Here are a few different questions and considerations.

First, please note that a load of 18 is not necessarily bad as long as you have at least 18 cores. How many actual CPU cores do you have?

Next, looking back at your original description you say `The Manager Node & Search node has 12 cores` and `The sensor node has 6 cores` so that's a total of 30 cores that you've assigned to VMs, right? 

Later you say `I did different combinations of heap size, CPU, total memory etc. but I am not able to lower the server load increasing to 16-18 within 2 days after I turn on the SO VMs . The SO VMs seems to utilize only 50-60% CPU on average.`

If your SO VMs are using 50% CPU and there's a total of 30 CPU cores assigned, then that would be roughly equal to a load of 15, right?

Is there an individual VM that is showing high load? If the high load is only on the Proxmox host, then perhaps the host resources are over-committed? One thing in particular to check for is swapping. Ideally, you would want to have more actual RAM than you've assigned to guests so that you can disable swap altogether.

Is there IO contention between the search node writing logs into Elasticsearch and the sensor node writing full packet capture to disk? You might try disabling full packet capture temporarily to see if that helps.

**Reply by Crimson1110** [reply]
It has total 2x Socket 8 cores so in total 16 cores. 32 threads.

Yes total of 30 Cores. I did not even spoke about the Sensor node in detail because it is consuming 5% of CPU and 10GB of 16Gb memory as the network is very idle.

As far as I know the load under Proxmox is sum of operations, including the CPU/IO/RAM. In any case, as you quoted, I did tests, like putting 8/20 cores to Manager/Search , less/more RAM etc. the load of the server did not change. I am going on with these tests more than a month and I am trying to understand how come one node consumes  full 8 threads (4 core) in a LAB environment with 20 idle machines (added to fleet for elastic agent which collects process.logs (total 5m entries last 24 hours). 

I even turned the event collection from Elastic agent off to see if it has an effect; it did not. 

No, all the other VM's are very idle.  Without SO the load is less than 1.  Now there are 10 Linux VMs and 2 Windows PCs are online.

For testing if the CPU is over committed, I separated the search node to another server with one CPU (same CPU as the other server E5-2630L v3). It has only the Search Node VM (12 CPU / 64gb RAM / 5x ssd Raid5) and average load is around 4 but it is increasing. The other server with all the VMs + Manager + Sensor node has a load of 10 and it is increasing. It takes 2-3 days to arrive to max / stabilize.

I disabled the Stenographer. It just lowered the writes on the sensor, but did not affect the overall load.

I even though perhaps the consumer level SSD's cache burned etc. but even if it is like that it should write at least 40mb/s and on InfluxDB / fio test I do not see anything unusual.  

Top on manager node shows user elastic+ Java 700% CPU...

**Reply by dougburks** [reply]
> One thing in particular to check for is swapping. Ideally, you would want to have more actual RAM than you've assigned to guests so that you can disable swap altogether.

Is there any swapping on your proxmox host or any guest?

> It has total 2x Socket 8 cores so in total 16 cores. 32 threads.

Are the VMs or processes pinned to physical sockets or are they floating back and forth?

> I even though perhaps the consumer level SSD's cache burned etc. but even if it is like that it should write at least 40mb/s and on InfluxDB / fio test I do not see anything unusual.

Might be worth running more diagnostics or trying different SSDs as there was a user that recently had an SSD causing performance issues:
https://github.com/Security-Onion-Solutions/securityonion/discussions/11813

**Reply by Crimson1110** [reply]
There is no swapping in any hosts or Proxmox. 

There are no pinned VMs. 

The idea to move the search node to entirely another server with new set of SSD disks was to eliminate that as well. But just to follow you I created a new RAID1 with 2 new SSD's and moved the Manager Node disk to the new LVM. Nothing changed.  Also it seems its averaging around 0.5Mb write which is nothing and if there was a problem with disks I would expect I/O in Influxdb would show up something. 

The problem could be related to Garbage Collection "young_collection_time_in_millis" as it keeps on growing (600k!). I do not know, it is getting complicated for me...

**Reply by dougburks** [reply]
> The problem could be related to Garbage Collection "young_collection_time_in_millis" as it keeps on growing (600k!). I do not know, it is getting complicated for me...

This isn't a setting that we commonly change. Which component are you referring to?

> There are no pinned VMs.

You might try pinning VMs to a particular socket to see if that makes any difference. In particular, you will most likely want to pin your sensor node to the same CPU that services your sniffing NIC:
https://docs.securityonion.net/en/2.4/zeek.html?highlight=numa#performance
https://docs.securityonion.net/en/2.4/suricata.html?highlight=numa#performance

Here's another idea to try. It seems like you're not monitoring that much network traffic and you're not consuming a huge amount of endpoint logs. So if you haven't already, you might try turning off all your existing Security Onion VMs and then creating a single VM, installing Security Onion in Standalone mode with default settings. See how that runs for a while and then try pinning that VM to the same CPU that services your sniffing NIC.

**Reply by Crimson1110** [reply]
> This isn't a setting that we commonly change. Which component are you referring to?

I was able to check "ParallelGCThreads" & "NewRatio" which seems ok to me. 

> Here's another idea to try. It seems like you're not monitoring that much network traffic and you're not consuming a huge amount of endpoint logs. So if you haven't already, you might try turning off all your existing Security Onion VMs and then creating a single VM, installing Security Onion in Standalone mode with default settings. See how that runs for a while and then try pinning that VM to the same CPU that services your sniffing NIC.

I was using standalone and decided to test the distributed version as I was hoping to introduce SO to some clients at some time. 

In the end, disabling the Playbook, where I enabled lots of rules, "kinda" did the trick. The search node was around 2.5 load on the dedicated  Proxmox host and on the other host with all the VMs + SO Manager + SO Sensor, the load dropped down to 5.  I moved the search node back to the host where others VMs are and the load is now around 9.4 - 9.8.  I hope it will stay like this.

Perhaps I will try to turn the Playbook again with fever rules enabled in the future. :(

Thank you very much for your help.

**Reply by dougburks** [reply, marked as answer]
> In the end, disabling the Playbook, where I enabled lots of rules, "kinda" did the trick. The search node was around 2.5 load on the dedicated Proxmox host and on the other host with all the VMs + SO Manager + SO Sensor, the load dropped down to 5. I moved the search node back to the host where others VMs are and the load is now around 9.4 - 9.8. I hope it will stay like this.

Yes, enabling lots of plays in Playbook will generate lots of Elasticsearch queries. Specifically, did you enable the `Malicious Nishang PowerShell Commandlets` play or any `experimental` plays?

From https://docs.securityonion.net/en/2.4/playbook.html#putting-a-play-into-production:
_Performance testing is still ongoing. We recommend avoiding the `Malicious Nishang PowerShell Commandlets` play as it can cause serious performance problems. You may also want to avoid others with a status of `experimental`._

**Reply by Crimson1110** [reply]
In the light of this conversation I suggest an edit to the "Read before posting" part:

- Did you read manual?
- Yes, at least 3 times.
- Read it again. 

Thanks a lot.  :)

I will enable only the rules that matter, and will be very careful about the ones that i'll write..  

![image](https://github.com/Security-Onion-Solutions/securityonion/assets/103431983/163b40bc-deb1-4237-bbc2-796ae9324fe6)

**Reply by dougburks** [reply]
I've updated the documentation and highlighted this as a WARNING:
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/1659467/41523c4b-eaac-4a7a-a8d8-b012e3d04a94)
